src/coreclr/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -69,8 +71,17 @@ public override bool Equals([NotNullWhen(true)] object? obj)
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern bool CanCompareBits(object obj);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern bool FastEqualsCheck(object a, object b);
+        private static unsafe bool FastEqualsCheck(object a, object b)
+        {
+            MethodTable* pMT = RuntimeHelpers.GetMethodTable(a);
+            Debug.Assert(!pMT->ContainsGCPointers);
+            Debug.Assert(!pMT->HasComponentSize);
+
+            // See RuntimeHelpers.GetRawObjectDataSize
+            // Omitting component size for value type
+            nuint rawSize = pMT->BaseSize - (nuint)(2 * sizeof(IntPtr));
```

Comment:
BaseSize includes padding. The original native code subtracts this padding to only compare the object payload.

Reply:
The padding is likely going to be all zeros, so the comparison is still going to work most of the time; but I am not sure whether the padding is guaranteed to be zero.

Reply:
Who knows maybe one day we decide to speed up boxing operation with `GC_ALLOC_ZEROING_OPTIONAL` 🙂 (at least for value types without paddings except the one in the header on 64bit)

Reply:
@jkotas Doesn't the `-2 * sizeof(IntPtr)` check above bypass the padding? This logic seems like a legal rewriting of the original C++ code to C#.

Reply:
@jkotas Doesn't the `-2 * sizeof(IntPtr)` check above bypass the padding? This logic seems like a legal rewriting of the original C++ code to C#.

Reply:
The existing code calls `TypeHandle::GetSize` that ends up calling `MethodTable::GetNumInstanceFieldBytes` that returns `GetBaseSize() - GetClass()->GetBaseSizePadding()`.

`GetClass()->GetBaseSizePadding()` is not always `-2 * sizeof(IntPtr)`

Reply:
Should we include corresponding field in `MethodTable` and use it instead?
Fixed size buffer can be a thing. I haven't test it.

Reply:
I've found that:

https://github.com/dotnet/runtime/blob/0dba0ee1b234100ade81d2b4f799cc118150785c/src/coreclr/vm/methodtablebuilder.cpp#L10507-L10509

OBJECT_BASESIZE is `2 * sizeof(IntPtr)`, MIN_OBJECT_SIZE is `3 * IntPtr`, and alignment is `sizeof(IntPtr)`.

So I'm getting confused now. The comment of `EEClass` is to share between generic instantiations, but `BaseSizePadding` can differ by instance fields.
Is the managed size calculation called by `object.MemberwiseClone()` really correct? It's probably safe because objects are always aligned.

Reply:
> Is the managed size calculation called by object.MemberwiseClone() really correct?

Yes, it is correct. `MemberwiseClone` does not need the exact number of bytes, it can copy the padding too without causing any harm.

Reply:
> Should we include corresponding field in MethodTable and use it instead?

I would add a new FCall that returns `TypeHandle::GetSize`.

Reply:
> Yes, it is correct. `MemberwiseClone` does not need the exact number of bytes, it can copy the padding too without causing any harm.

I think it's better to add a comment for it to prevent misuse.

Reply:
> I would add a new FCall that returns `TypeHandle::GetSize`.

I wonder if this PR is beneficial at all. FCall can't be inlined right? There's still 1 FCall.

Reply:
The new FCall would be simpler than the current one, so there is still some benefit. 

I am not too concerned about the performance of this path. ValueTypes that care about Equals/GetHashCode performance should override Equals/GetHashCode.